File: service/resource-encapsulation/src/resourceClient/RCSRemoteResourceObject.cpp
//******************************************************************
//
// Copyright 2015 Samsung Electronics All Rights Reserved.
//
//-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
//-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

#include "RCSRemoteResourceObject.h"
#include "ResourceBroker.h"
#include "ResourceCacheManager.h"

#define CLIENT_W_TAG  PCF("RCSRemoteResourceObject")

using namespace OIC::Service;

namespace
{
    ResourceState getResourceStateFromBrokerState(BROKER_STATE state)
    {

        OC_LOG(DEBUG, CLIENT_W_TAG, "getResourceStateFromBrokerState entry");

        if (state == BROKER_STATE::ALIVE)
        {
            return ResourceState::ALIVE;
        }
        else if (state == BROKER_STATE::REQUESTED)
        {
            return ResourceState::REQUESTED;
        }
        else if (state == BROKER_STATE::LOST_SIGNAL)
        {
            return ResourceState::LOST_SIGNAL;
        }
        else if (state == BROKER_STATE::DESTROYED)
        {
            return ResourceState::DESTROYED;
        }
        else if (state == BROKER_STATE::NONE)
        {
            return ResourceState::NOT_MONITORING;
        }
        OC_LOG(ERROR, CLIENT_W_TAG, "getResourceStateFromBrokerState ERROR");

        //Default return value
        return ResourceState::NOT_MONITORING;
    }

    CacheState getCacheState(CACHE_STATE state)
    {

        OC_LOG(DEBUG, CLIENT_W_TAG, "getCacheState (from CACHE_STATE) entry");

        if (state == CACHE_STATE::READY)
        {
            return CacheState::READY;
        }
        else if (state == CACHE_STATE::READY_YET)
        {
            return CacheState::READY_YET;
        }
        else if (state == CACHE_STATE::LOST_SIGNAL)
        {
            return CacheState::LOST_SIGNAL;
        }
        else if (state == CACHE_STATE::DESTROYED)
        {
            return CacheState::DESTROYED;
        }
        else if (state == CACHE_STATE::UPDATING)
        {
            return CacheState::UPDATING;
        }
        else if (state == CACHE_STATE::NONE)
        {
            return CacheState::NONE;
        }
        OC_LOG(ERROR, CLIENT_W_TAG, "getCacheState (from CACHE_STATE) ERROR");

        //Default return value
        return CacheState::NONE;
    }

    OCStackResult hostingCallback(BROKER_STATE state,
                                  RCSRemoteResourceObject::ResourceStateChangedCallback onResourceStateChanged)
    {
        OC_LOG(DEBUG, CLIENT_W_TAG, "hostingCallback entry");

        ResourceState resourceState = getResourceStateFromBrokerState(state);
        onResourceStateChanged(resourceState); //passing ResourceState to application

        OC_LOG(DEBUG, CLIENT_W_TAG, "hostingCallback exit");
        return OC_STACK_OK;
    }

    OCStackResult cachingCallback(std::shared_ptr<PrimitiveResource> resource,
                                  const RCSResourceAttributes &data,
                                  RCSRemoteResourceObject::CacheUpdatedCallback onCacheUpdated)
    {
        OC_LOG(DEBUG, CLIENT_W_TAG, "cachingCallback entry");

        onCacheUpdated(data); //passing ResourceAttributes to application

        OC_LOG(DEBUG, CLIENT_W_TAG, "cachingCallback exit");
        return OC_STACK_OK;
    }

    void setCallback(const HeaderOptions &header, const ResponseStatement &response, int n,
                     RCSRemoteResourceObject::RemoteAttributesSetCallback onRemoteAttributesSet)
    {
        OC_LOG(DEBUG, CLIENT_W_TAG, "setCallback entry");

        const RCSResourceAttributes &attributes = response.getAttributes();
        onRemoteAttributesSet(attributes); //passing ResourceAttributes to application

        OC_LOG(DEBUG, CLIENT_W_TAG, "setCallback exit");
    }

    void getCallback(const HeaderOptions &headerOption, const ResponseStatement &response, int n,
                     RCSRemoteResourceObject::RemoteAttributesReceivedCallback onRemoteAttributesReceived)
    {
        OC_LOG(DEBUG, CLIENT_W_TAG, "getCallback entry");

        const RCSResourceAttributes &attributes = response.getAttributes();
        onRemoteAttributesReceived(attributes); //passing ResourceAttributes to application

        OC_LOG(DEBUG, CLIENT_W_TAG, "getCallback exit");
    }
}

//******************************* RCSRemoteResourceObject *************************************

namespace OIC
{
    namespace Service
    {
        RCSRemoteResourceObject:: RCSRemoteResourceObject(std::shared_ptr<PrimitiveResource>  pResource) :
            m_monitoringFlag(false), m_cachingFlag(false),  m_observableFlag(pResource->isObservable()),
            m_primitiveResource(pResource), m_cacheId(0), m_brokerId(0) {}

        bool RCSRemoteResourceObject::isMonitoring() const
        {
            return m_monitoringFlag;
        }

        bool RCSRemoteResourceObject::isCaching() const
        {
            return m_cachingFlag;
        }

        void RCSRemoteResourceObject::startMonitoring(ResourceStateChangedCallback cb)
        {
            OC_LOG(DEBUG, CLIENT_W_TAG, "RCSRemoteResourceObject::startMonitoring entry");
            if (true == m_monitoringFlag)
            {
                OC_LOG(DEBUG, CLIENT_W_TAG, "RCSRemoteResourceObject::startMonitoring : Already started");
            }
            else
            {
                try
                {
                    BrokerID brokerId =  ResourceBroker::getInstance()->hostResource(m_primitiveResource,
                                         std::bind(hostingCallback, std::placeholders::_1,
                                                   cb));
                    m_monitoringFlag = true;
                    m_brokerId = brokerId;
                }
                catch (std::exception &exception)
                {
                    throw InvalidParameterException {exception.what()};
                }
            }
            OC_LOG(DEBUG, CLIENT_W_TAG, "RCSRemoteResourceObject::startMonitoring exit");
        }

        void RCSRemoteResourceObject::stopMonitoring()
        {
            OC_LOG(DEBUG, CLIENT_W_TAG, "RCSRemoteResourceObject::stopMonitoring entry");
            if (true == m_monitoringFlag)
            {
                try
                {
                    ResourceBroker::getInstance()->cancelHostResource(m_brokerId);
                    m_monitoringFlag = false;
                }
                catch (std::exception &exception)
                {
                    OC_LOG(DEBUG, CLIENT_W_TAG, "RCSRemoteResourceObject::stopMonitoring InvalidParameterException");
                }
            }
            else
            {
                OC_LOG(DEBUG, CLIENT_W_TAG, "RCSRemoteResourceObject:: stopMonitoring : already terminated");
            }

            OC_LOG(DEBUG, CLIENT_W_TAG, "RCSRemoteResourceObject::stopMonitoring exit");
        }

        ResourceState RCSRemoteResourceObject::getState() const
        {
            OC_LOG(DEBUG, CLIENT_W_TAG, " RCSRemoteResourceObject::getState entry");
            try
            {
                BROKER_STATE brokerState = ResourceBroker::getInstance()->getResourceState(m_primitiveResource);
                OC_LOG(DEBUG, CLIENT_W_TAG, " RCSRemoteResourceObject::getState exit");
                return getResourceStateFromBrokerState(brokerState);
            }
            catch (std::exception &exception)
            {
                OC_LOG(DEBUG, CLIENT_W_TAG, " RCSRemoteResourceObject::getState InvalidParameterException");
                throw BadRequestException { "[getState] Get Resource Source State from Broker Error " };
            }
        }

        void RCSRemoteResourceObject::startCaching()
        {
            OC_LOG(DEBUG, CLIENT_W_TAG, "RCSRemoteResourceObject::startCaching entry");
            if (true == m_cachingFlag)
            {
                OC_LOG(DEBUG, CLIENT_W_TAG, "RCSRemoteResourceObject::startCaching : already Started");
            }
            else
            {
                try
                {
                    CacheID cacheId = ResourceCacheManager::getInstance()->requestResourceCache(m_primitiveResource,
                                      NULL, REPORT_FREQUENCY::NONE,  0);

                    m_cacheId = cacheId;
                    m_cachingFlag = true;
                    OC_LOG_V(DEBUG, CLIENT_W_TAG, "RCSRemoteResourceObject::startCaching CACHE ID %d", cacheId);
                }
                catch (std::exception &exception)
                {
                    OC_LOG(DEBUG, CLIENT_W_TAG, "RCSRemoteResourceObject::startCaching InvalidParameterException");
                }
            }
            OC_LOG(DEBUG, CLIENT_W_TAG, "RCSRemoteResourceObject::startCaching exit");
        }

        void RCSRemoteResourceObject::startCaching(CacheUpdatedCallback cb)
        {
            OC_LOG(DEBUG, CLIENT_W_TAG, "RCSRemoteResourceObject::startCaching entry");
            if (!cb)
            {
                throw InvalidParameterException {"startCaching : Callback is NULL" };
            }
            else
            {
                if (true == m_cachingFlag)
                {
                    OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::startCaching : already Started");
                }
                else
                {
                    try
                    {
                        CacheID cacheId = ResourceCacheManager::getInstance()->requestResourceCache(m_primitiveResource,
                                          std::bind(cachingCallback, std::placeholders::_1, std::placeholders::_2, cb),
                                          REPORT_FREQUENCY::UPTODATE,  0);

                        m_cacheId = cacheId;
                        m_cachingFlag = true;
                        OC_LOG_V(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::startCaching CACHE ID %d", cacheId);
                    }
                    catch (std::exception &exception)
                    {
                        throw InvalidParameterException {"startCaching : error" };
                    }
                }
            }
        }

        void RCSRemoteResourceObject::stopCaching()
        {
            OC_LOG(DEBUG, CLIENT_W_TAG, "RCSRemoteResourceObject::stopCaching entry");

            if (true == m_cachingFlag)
            {
                try
                {
                    ResourceCacheManager::getInstance()->cancelResourceCache(m_cacheId);
                    m_cachingFlag = false;
                }
                catch (std::exception &exception)
                {
                    OC_LOG(DEBUG, CLIENT_W_TAG, "RCSRemoteResourceObject::stopCaching InvalidParameterException");
                }
            }
            else
            {
                OC_LOG(DEBUG, CLIENT_W_TAG, "RCSRemoteResourceObject:: Caching already terminated");
            }

            OC_LOG(DEBUG, CLIENT_W_TAG, "RCSRemoteResourceObject::stopCaching exit");
        }

        CacheState  RCSRemoteResourceObject::getResourceCacheState()
        {
            OC_LOG(DEBUG, CLIENT_W_TAG, "RCSRemoteResourceObject::getResourceCacheState entry");
            try
            {
                CACHE_STATE cacheState = ResourceCacheManager::getInstance()->getResourceCacheState(
                                             m_primitiveResource);
                OC_LOG(DEBUG, CLIENT_W_TAG, "RCSRemoteResourceObject::getResourceCacheState exit");
                return getCacheState(cacheState);
            }
            catch (std::exception &exception)
            {
                OC_LOG(DEBUG, CLIENT_W_TAG,
                       "RCSRemoteResourceObject::getResourceCacheState InvalidParameterException");
                throw BadRequestException { "[getResourceCacheState] Caching not started" };
            }
        }

        void RCSRemoteResourceObject::refreshCache()
        {
            OC_LOG(DEBUG, CLIENT_W_TAG, "RCSRemoteResourceObject::refreshCache entry");
            try
            {

                ResourceCacheManager::getInstance()->updateResourceCache(
                    m_primitiveResource);
            }
            catch (std::exception &exception)
            {
                OC_LOG(DEBUG, CLIENT_W_TAG, "RCSRemoteResourceObject::refreshCache InvalidParameterException");
            }
            OC_LOG(DEBUG, CLIENT_W_TAG, "RCSRemoteResourceObject::refreshCache exit");
        }

        RCSResourceAttributes RCSRemoteResourceObject:: getCachedAttributes() const
        {
            OC_LOG(DEBUG, CLIENT_W_TAG, "RCSRemoteResourceObject :: getCachedAttributes ");
            try
            {
                return ResourceCacheManager::getInstance()->getCachedData(m_primitiveResource);
            }
            catch (std::exception &exception)
            {
                throw BadRequestException { "[getCachedAttributes]  Caching not started" };
            }
        }

        RCSResourceAttributes::Value  RCSRemoteResourceObject:: getCachedAttribute( const std::string &key)
        {
            OC_LOG(DEBUG, CLIENT_W_TAG, "RCSRemoteResourceObject :: getCachedAttribute entry");
            try
            {
                RCSResourceAttributes  Cachedattributes = ResourceCacheManager::getInstance()->getCachedData(
                            m_primitiveResource);
                return Cachedattributes[key];
            }
            catch (std::exception &exception)
            {
                throw BadRequestException { "[getCachedAttribute]  Caching not started" };
            }
            OC_LOG(DEBUG, CLIENT_W_TAG, "RCSRemoteResourceObject :: getCachedAttribute exit");
        }

        std::string RCSRemoteResourceObject::getUri() const
        {
            return m_primitiveResource->getUri();
        }

        std::string RCSRemoteResourceObject::getAddress() const
        {
            return m_primitiveResource->getHost();
        }

        std::vector < std::string > RCSRemoteResourceObject::getTypes() const
        {
            return m_primitiveResource->getTypes();
        }

        std::vector < std::string > RCSRemoteResourceObject::getInterfaces() const
        {
            return m_primitiveResource->getInterfaces();
        }

        void RCSRemoteResourceObject::getRemoteAttributes(RemoteAttributesReceivedCallback cb)
        {
            OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::getRemoteAttributes entry");
            if (!cb)
            {
                throw InvalidParameterException {"getRemoteAttributes : Callback is NULL" };
            }
            else
            {
                m_primitiveResource->requestGet(std::bind(getCallback, std::placeholders::_1,
                                                std::placeholders::_2, std::placeholders::_3, cb));
            }
            OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::getRemoteAttributes exit");
        }

        void RCSRemoteResourceObject::setRemoteAttributes(const RCSResourceAttributes &attribute,
                RemoteAttributesSetCallback cb)
        {
            OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::setRemoteAttributes entry");
            if (!cb)
            {
                throw InvalidParameterException {"setRemoteAttributes : Callback is NULL" };
            }
            else
            {
                m_primitiveResource->requestSet(attribute, std::bind(setCallback, std::placeholders::_1,
                                                std::placeholders::_2, std::placeholders::_3, cb));
            }
            OC_LOG(DEBUG, CLIENT_W_TAG, "RemoteResourceObject::setRemoteAttributes exit");
        }


        bool RCSRemoteResourceObject::isObservable() const
        {
            return m_observableFlag;
        }
    }
}

File: service/resource-encapsulation/sdk/java/src/org/iotivity/ResourceEncapsulation/client/RCSRemoteResourceObject.java
/******************************************************************
 *
 * Copyright 2015 Samsung Electronics All Rights Reserved.
 *
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 ******************************************************************/
/**
 * @file    RCSRemoteResourceObject.java
 *
 * @brief     This file provides a class which provides a set of native functions and API's
 *        of resource client, provided to the developers.
 */

package org.iotivity.ResourceEncapsulation.client;

import java.util.List;
import java.util.Vector;

import org.iotivity.ResourceEncapsulation.common.RCSException;
import org.iotivity.ResourceEncapsulation.common.RCSResourceAttributes;

import android.util.Log;

/**
 *
 * The resource can be discovered with discoverResource. On successful resource
 * discovery an instance of this class is returned. This class is an interaction
 * point between Resource and the developers. Developer will get the
 * RCSRemoteResourceObject by calling RCSDiscoveryManager's discoverResource
 * API.
 *
 * {@link RCSDiscoveryManager}
 *
 */
public class RCSRemoteResourceObject {

    private IStateChangedCallbackListener m_stateChangedListener;
    private ICacheUpdateListener          m_cacheUpdateListener;
    private IRemoteAttributeListener      m_remoteAttributeListener;
    private long                          m_nativeHandle;
    private String                        LOG_TAG = this.getClass()
                                                          .getSimpleName();

    // native methods
    private native boolean nativeIsMonitoring();

    private native boolean nativeIsCaching();

    private native boolean nativeIsObservable();

    private native void nativeStartMonitoring(
            IStateChangedCallbackListener listener);

    private native void nativeStopMonitoring();

    private native int nativeGetState();

    private native void nativeStartCaching(ICacheUpdateListener listener);

    private native void nativeStopCaching();

    private native int nativeGetCacheState();

    private native boolean nativeIsCachedAvailable();

    private native RCSResourceAttributes nativeGetCachedAttributes();

    private native String nativeGetCachedAttribute(String key);

    private native void nativeGetRemoteAttributes(
            IRemoteAttributeListener listener);

    private native void nativeSetRemoteAttributes(
            RCSResourceAttributes attributes, IRemoteAttributeListener listener);

    private native String nativeGetUri();

    private native String nativeGetAddress();

    private native List<String> nativeGetTypes();

    private native List<String> nativeGetInterFaces();

    private RCSRemoteResourceObject(long nativeHandle) {
        this.m_nativeHandle = nativeHandle;
    }

    /**
     * This Enum contains Status codes for resource state.
     *
     * <p>
     * Developer will get the ResoureState as a result of getState API of
     * RCSRemoteResourceObject class and as a callback result of startMonitoring
     * API
     *
     * {@link RCSRemoteResourceObject}
     */
    public enum ResourceState {

        NONE, /** < Monitoring is not started. */
        REQUESTED, /**
         * < Monitoring is started and checking state is in progress.
         * This is the default state after startMonitoring.
         */
        ALIVE, /** < The resource is alive. */
        LOST_SIGNAL, /** < Failed to reach the resource. */
        DESTROYED;
        /** < The resource is deleted. */

        /**
         * Utility API for integer to enum conversion.
         */
        public static ResourceState getResourceState(int ordinal) {

            if (ordinal == 0)
                return ResourceState.values()[0];
            else if (ordinal == 1)
                return ResourceState.values()[1];
            else if (ordinal == 2)
                return ResourceState.values()[2];
            else if (ordinal == 3)
                return ResourceState.values()[3];

            return null;
        }
    }

    /**
     * This Enum contains Status codes for cache state.
     *
     * <p>
     * Application will get this CacheState as a result of getCacheState API of
     * RCSRemoteResourceObject class
     *
     * {@link RCSRemoteResourceObject}
     */
    public enum CacheState {

        NONE, /** < Caching is not started. */
        UNREADY, /**
         * < Caching is started, but the data is not ready yet. This is
         * the default state after startCaching.
         */
        READY, /** < The data is ready. */
        LOST_SIGNAL;
        /** < Failed to reach the resource. */

        /**
         * Utility API for int to enum conversion.
         */
        public static CacheState getCacheState(int ordinal) {

            CacheState result = CacheState.values()[0];

            if (ordinal == 1)
                result = CacheState.values()[1];
            else if (ordinal == 2)
                result = CacheState.values()[2];
            else if (ordinal == 3)
                result = CacheState.values()[3];

            return result;
        }
    }

    /**
     * Provides interface for receiving the callback on cache update.
     */
    public interface ICacheUpdateListener {

        /**
         * This callback method is called when a asynchronous response for
         * startCaching request is received. Listener should be set to receive
         * this callback.
         *
         * @param attributes
         *            It is the updated resource attributes.
         *
         *            {@link RCSResourceAttributes}
         */
        public void onCacheUpdatedCallback(RCSResourceAttributes attributes);

    }

    /**
     * Provides interface for receiving the callback for the getRemoteAttributes
     * and setRemoteAttributes requested actions.
     */
    public interface IRemoteAttributeListener {

        /**
         * This callback method is called when a asynchronous response for the
         * getRemoteAttributes request or setRemoteAttributes request is
         * received.
         *
         * @param attributes
         *            It is the updated resource attributes value received from
         *            the remote resource.
         *
         *            {@link RCSResourceAttributes}
         */
        public void onRemoteAttributesGetCallback(
                RCSResourceAttributes attributes);

    }

    /**
     * Provides interface for receiving the callback for the startMonitoring
     * This callback is received on remote resource state change.
     */
    public interface IStateChangedCallbackListener {

        /**
         * This callback method is called when a asynchronous response for the
         * startMonitoring request is received.
         *
         * @param resourceState
         *            It gives the current remote resource state.
         *
         *            {@link ResourceState}
         */
        public void onStateChanged(ResourceState resourceState);
    }

    /**
     * Set listener for receiving asynchronous response for cache updates.
     *
     * @param listener
     *            ICacheUpdateListener to receive asynchronous response for
     *            cache updates.
     */
    public void setCacheUpdateListener(ICacheUpdateListener listener) {
        m_cacheUpdateListener = listener;
    }

    /**
     * Set listener for receiving asynchronous response for remote resource
     * state changes.
     *
     * @param listener
     *            IStateChangedCallbackListener to receive asynchronous response
     *            for remote resource state changes.
     */
    public void setStateChangedListener(IStateChangedCallbackListener listener) {
        m_stateChangedListener = listener;
    }

    /**
     * Set listener for receiving asynchronous response for resource attributes
     * SET and GET request.
     *
     * @param listener
     *            IRemoteAttributeListener to receive asynchronous response for
     *            resource attributes SET and GET request.
     */
    public void setRemoteAttributesListener(IRemoteAttributeListener listener) {
        m_remoteAttributeListener = listener;
    }

    /**
     * Returns whether monitoring is enabled.
     *
     * @return boolean - true if monitoring the resource.
     *
     *         {@link #startMonitoring()}
     */
    public boolean isMonitoring() {

        Log.i(LOG_TAG, "isMonitoring called");
        return this.nativeIsMonitoring();
    }

    /**
     * Returns whether caching is enabled.
     *
     * @return boolean - true if caching the resource.
     *
     *         {@link #startCaching()}
     */
    public boolean isCaching() {

        Log.i(LOG_TAG, "isCaching called");
        return this.nativeIsCaching();
    }

    /**
     * Returns whether resource is observable.
     *
     * @return boolean - true if resource is observable.
     */
    public boolean isObservable() {

        Log.i(LOG_TAG, "isObservable called");
        return this.nativeIsObservable();
    }

    /**
     * Starts monitoring the resource.
     * <p>
     * Monitoring provides a feature to check the presence of a resource, even
     * when the server is not announcing Presence using startPresnece.
     * <p>
     * StateChangedListener should be set before calling this API
     *
     * @throws RCSException
     *             If monitoring is already started.
     *
     *             {@link IStateChangedCallbackListener} {@link ResourceState}
     *             {@link #isMonitoring()} {@link #stopMonitoring()}
     */
    public void startMonitoring() throws RCSException {

        Log.i(LOG_TAG, "startMonitoring called");
        this.nativeStartMonitoring(m_stateChangedListener);
    }

    /**
     * Stop monitoring the resource.
     * <p>
     * It does nothing if monitoring is not started.
     * <p>
     * {@link #startMonitoring()}
     */
    public void stopMonitoring() {

        Log.i(LOG_TAG, "stopMonitoring called");
        this.nativeStopMonitoring();
    }

    /**
     * Returns the current state of the resource.
     *
     * @return ResourceState - current resource state
     *
     *         {@link #startMonitoring()} {@link ResourceState}
     */
    public ResourceState getState() {

        Log.i(LOG_TAG, "getState called");
        int state = this.nativeGetState();

        return ResourceState.getResourceState(state);
    }

    /**
     * Starts caching attributes of the resource.
     *
     * This will start data caching for the resource. Once caching started it
     * will look for the data updation on the resource and updates the cache
     * data accordingly.
     *
     * @throws RCSException
     *
     *             <p>
     *             If this method is called with null listener it will not
     *             notify the updated value to caller. It it is called with a
     *             valid listener updated cached values will be notified to
     *             caller.
     *
     *             {@link #getCacheState()} {@link #getCachedAttributes()}
     */
    public void startCaching() throws RCSException {

        if (null == m_cacheUpdateListener) {
            Log.i(LOG_TAG, "startCaching WITHOUT callback");
        } else {
            Log.i(LOG_TAG, "startCaching WITH callback");
        }

        this.nativeStartCaching(m_cacheUpdateListener);
    }

    /**
     * Stops caching.
     *
     * It does nothing if caching is not started.
     *
     * {@link #startCaching()}
     */
    public void stopCaching() {

        Log.i(LOG_TAG, "stopCaching called");
        this.nativeStopCaching();
    }

    /**
     * Returns the current cache state.
     *
     * @return CacheState - current cache state.
     *
     *         {@link CacheState}
     */
    public CacheState getCacheState() {

        Log.i(LOG_TAG, "getCacheState called");
        return CacheState.getCacheState(this.nativeGetCacheState());
    }

    /**
     * Returns whether cached data is available.
     *
     * Cache will be available always after cache state READY even if current
     * state is LOST_SIGNAL.
     *
     * @return boolean - true if cache data is available.
     *
     *         {@link #startCaching()} {@link CacheState}
     */
    public boolean isCachedAvailable() {

        Log.i(LOG_TAG, "isCachedAvailable called");
        return this.nativeIsCachedAvailable();
    }

    /**
     * Gets the cached RCSResourceAttributes data.
     *
     * <p>
     * Cache should be available.
     *
     * @return RCSResourceAttributes
     *
     * @throws RCSException
     *             If the Cache is not started.
     *
     *             {@link RCSResourceAttributes} {@link #isCachedAvailable()}
     *             {@link #startCaching()}
     */
    public RCSResourceAttributes getCachedAttributes() throws RCSException {

        Log.i(LOG_TAG, "getCachedAttributes called");
        return this.nativeGetCachedAttributes();
    }

    /**
     * Gets a particular cached ResourceAttribute Value.
     *
     * <p>
     * Cache should be available.
     *
     * @param key
     *            particular attribute name as a string to get the value for.
     *
     * @return String - Requested attribute value in string.
     *
     * @throws RCSException
     *             If the Cache is not started or InvalidKeyException If @a key
     *             doesn't match the key of any value.
     *
     *             {@link RCSResourceAttributes} {@link #isCachedAvailable()}
     *             {@link #startCaching()}
     */
    public String getCachedAttribute(String key) throws RCSException {

        Log.i(LOG_TAG, "getCachedAttribute called");
        return this.nativeGetCachedAttribute(key);
    }

    /**
     * Gets resource attributes directly from the server.
     *
     * This API send a get request to the resource of interest and provides the
     * attributes to the caller in the RemoteAttributesReceivedCallback.
     *
     * <p>
     * IRemoteAttributeListener should be set to get the attributes
     *
     * {@link IRemoteAttributeListener}
     */
    public void getRemoteAttributes() throws RCSException {

        Log.i(LOG_TAG, "getRemoteAttributes called");
        this.nativeGetRemoteAttributes(m_remoteAttributeListener);
    }

    /**
     * Sends a set request with resource attributes to the server.
     *
     * The SetRequest behavior depends on the server, whether updating its
     * attributes or not.
     *
     * @param attributes
     *            Attribute value to set for the remote resource.
     *
     *            <p>
     *            IRemoteAttributeListener should be set to call this API
     *
     *            {@link IRemoteAttributeListener}
     */
    public void setRemoteAttributes(RCSResourceAttributes attributes)
            throws RCSException {

        Log.i(LOG_TAG, "getRemoteAttributes called");
        this.nativeSetRemoteAttributes(attributes, m_remoteAttributeListener);
    }

    /**
     * Returns the uri of the resource.
     *
     * @return Uri of the resource
     */
    public String getUri() {

        Log.i(LOG_TAG, "getUri called");
        return this.nativeGetUri();
    }

    /**
     * Returns the address of the resource .
     *
     * @return Address of the resource
     */
    public String getAddress() {

        Log.i(LOG_TAG, "getAddress called");
        return this.nativeGetAddress();
    }

    /**
     * Returns the resource types of the resource.
     *
     * @return vector containing resource types
     */
    public Vector<String> getTypes() {

        Log.i(LOG_TAG, "getTypes called");
        List<String> types = this.nativeGetTypes();
        Vector<String> resourceType = new Vector<String>();
        resourceType.addAll(types);
        return resourceType;
    }

    /**
     * Returns the resource interfaces of the resource.
     *
     * @return vector containing resource Interfaces
     */
    public Vector<String> getInterFaces() {

        Log.i(LOG_TAG, "getInterFaces called");
        List<String> interfaces = this.nativeGetInterFaces();
        Vector<String> resourceInterfaces = new Vector<String>();
        resourceInterfaces.addAll(interfaces);
        return resourceInterfaces;
    }
}
